src/main/java/com/microsoft/sqlserver/jdbc/SQLServerPreparedStatement.java

```diff
@@ -2103,23 +2103,26 @@ public int[] executeBatch() throws SQLServerException, BatchUpdateException, SQL
                         batchRecord.addColumnMetadata(i, c.getColumnName(), jdbctype, ti.getPrecision(), ti.getScale());
                     }
 
-                    SQLServerBulkCopy bcOperation = new SQLServerBulkCopy(connection);
-                    SQLServerBulkCopyOptions option = new SQLServerBulkCopyOptions();
-                    option.setBulkCopyTimeout(queryTimeout);
-                    bcOperation.setBulkCopyOptions(option);
-                    bcOperation.setDestinationTableName(tableName);
-                    bcOperation.setStmtColumnEncriptionSetting(this.getStmtColumnEncriptionSetting());
-                    bcOperation.setDestinationTableMetadata(rs);
-                    bcOperation.writeToServer(batchRecord);
-                    bcOperation.close();
-                    updateCounts = new int[batchParamValues.size()];
-                    for (int i = 0; i < batchParamValues.size(); ++i) {
-                        updateCounts[i] = 1;
-                    }
+                    try {
+                        SQLServerBulkCopy bcOperation = new SQLServerBulkCopy(connection);
+                        SQLServerBulkCopyOptions option = new SQLServerBulkCopyOptions();
+                        option.setBulkCopyTimeout(queryTimeout);
+                        bcOperation.setBulkCopyOptions(option);
+                        bcOperation.setDestinationTableName(tableName);
+                        bcOperation.setStmtColumnEncriptionSetting(this.getStmtColumnEncriptionSetting());
+                        bcOperation.setDestinationTableMetadata(rs);
+                        bcOperation.writeToServer(batchRecord);
+                        bcOperation.close();
+                        updateCounts = new int[batchParamValues.size()];
+                        for (int i = 0; i < batchParamValues.size(); ++i) {
+                            updateCounts[i] = 1;
+                        }
 
-                    batchParamValues = null;
-                    loggerExternal.exiting(getClassNameLogging(), "executeBatch", updateCounts);
-                    return updateCounts;
+                        loggerExternal.exiting(getClassNameLogging(), "executeBatch", updateCounts);
+                        return updateCounts;
+                    } finally {
```

Comment:
This can be moved to 2133, `batchParamValues = null` should be executed in all cases so it makes sense for it to be in a `finally` block belonging to the initial `try` block.

Reply:
Alternative approach could be to remove both the additional inner `try` I've added here _and_ the one at line 2296 in the "old style batch insert" branch, and instead wrap the whole method body in an outer try/finally that ensures `batchParamValues` is cleared in all cases immediately before the method returns.  That might be clearer overall.

Reply:
I've made a separate PR #1869 for the alternative approach, which is also rebased against the latest main.